File: DSA Implementations/C++/depth_first_search.cpp
#include <iostream>
#include <vector>
#include <stack>
#include <list>

using namespace std;

// Graph class
class Graph
{
private:
    int numVertices;
    bool weighted;
    vector<list<pair<int, int>>> adjLists; // adjacency list (pair(vertex, weight))

public:
    Graph(int vertices, bool isWeighted)
    {
        numVertices = vertices;
        weighted = isWeighted;
        adjLists.resize(vertices); // resize vector to hold adjacency lists for each vertex
    }

    // Function to add a directed edge
    void addSingleEdge(int src, int dest, int weight)
    {
        adjLists[src].emplace_back(dest, weighted ? weight : 1); // add edge with weight or default to 1
    }

    // Function to add an edge (handles directed/undirected graph)
    void addEdge(int src, int dest, bool directed, int weight)
    {
        addSingleEdge(src, dest, weight); // add directed edge
        if (!directed)
        {
            addSingleEdge(dest, src, weight); // add the reverse edge for undirected graph
        }
    }

    // Function to print the graph (adjacency list)
    void printGraph()
    {
        cout << "Adjacency List Representation: " << endl;
        for (int v = 0; v < numVertices; v++)
        {
            cout << v << " : ";
            for (auto &neighbor : adjLists[v])
            {
                if (weighted)
                {
                    cout << neighbor.first << "(" << neighbor.second << ") ";
                }
                else
                {
                    cout << neighbor.first << " ";
                }
            }
            cout << endl;
        }
        cout << endl;
    }

    // DFS Traversal function
    void DFS(int startVertex)
    {
        vector<bool> visited(numVertices, false);
        stack<int> frontier;

        frontier.push(startVertex);
        cout << "DFS Traversal starting from vertex " << startVertex << ": " << endl;

        while (!frontier.empty())
        {
            int current = frontier.top();
            frontier.pop();

            if (!visited[current])
            {
                visited[current] = true;
                cout << current << " ";

                for (auto &neighbor : adjLists[current])
                {
                    if (!visited[neighbor.first])
                    {
                        frontier.push(neighbor.first);
                    }
                }
            }
        }
        cout << endl;
    }
};

int main()
{
    int vertices, edges, directed, weighted;

    cout << "Enter the number of vertices (Maximum 100): ";
    cin >> vertices;

    cout << "Enter the number of edges: ";
    cin >> edges;

    cout << "Is the graph directed? (0 = No, 1 = Yes): ";
    cin >> directed;

    cout << "Is the graph weighted? (0 = No, 1 = Yes): ";
    cin >> weighted;

    Graph graph(vertices, weighted);

    cout << "Enter the edges one by one:- " << endl;

    for (int i = 0; i < edges; i++)
    {
        int s, d, w = 1;
        cout << "Edge " << i + 1 << ": ";
        if (weighted)
        {
            cin >> s >> d >> w;
        }
        else
        {
            cin >> s >> d;
        }
        graph.addEdge(s, d, directed, w);
    }

    int startVertex;
    cout << "Enter the starting vertex for DFS: ";
    cin >> startVertex;

    graph.printGraph();     // print the adjacency list
    graph.DFS(startVertex); // perform DFS traversal

    return 0;
}

File: DSA Implementations/Python/depth_first_search.py
class Graph:
    def __init__(self, vertices, weighted):
        self.numVertices = vertices
        self.weighted = weighted
        self.adjLists = [[] for _ in range(vertices)]  # adjacency list

    def add_single_edge(self, src, dest, weight):
        self.adjLists[src].append((dest, weight if self.weighted else 1))  # add edge with weight or default to 1

    def add_edge(self, src, dest, directed, weight):
        self.add_single_edge(src, dest, weight)  # add directed edge
        if not directed:
            self.add_single_edge(dest, src, weight)  # add the reverse edge for undirected graph

    def print_graph(self):
        print("Adjacency List Representation: ")
        for v in range(self.numVertices):
            print(f"{v} : ", end="")
            for neighbor in self.adjLists[v]:
                if self.weighted:
                    print(f"{neighbor[0]}({neighbor[1]}) ", end="")
                else:
                    print(f"{neighbor[0]} ", end="")
            print()

    def dfs(self, start_vertex):
        visited = [False] * self.numVertices
        stack = [start_vertex]
        print(f"DFS Traversal starting from vertex {start_vertex}: ")
        
        while stack:
            current = stack.pop()

            if not visited[current]:
                visited[current] = True
                print(current, end=" ")

                for neighbor in self.adjLists[current]:
                    if not visited[neighbor[0]]:
                        stack.append(neighbor[0])

        print()

# Main code
if __name__ == "__main__":
    vertices = int(input("Enter the number of vertices (Maximum 100): "))
    edges = int(input("Enter the number of edges: "))
    directed = int(input("Is the graph directed? (0 = No, 1 = Yes): "))
    weighted = int(input("Is the graph weighted? (0 = No, 1 = Yes): "))

    graph = Graph(vertices, weighted)

    print("Enter the edges one by one:- ")

    for i in range(edges):
        if weighted:
            s, d, w = map(int, input(f"Edge {i + 1}: ").split())
        else:
            s, d = map(int, input(f"Edge {i + 1}: ").split())
            w = 1  # default weight
        graph.add_edge(s, d, directed, w)

    start_vertex = int(input("Enter the starting vertex for DFS: "))

    graph.print_graph()  # print the adjacency list
    graph.dfs(start_vertex)  # perform DFS traversal